### ISSUE TITLE
Remove usage of `docker` in Gardener, such that Gardener can use OS images without docker daemon installed

### ISSUE THREAD

**Issue by voelzmo** [gardener/gardener#4673]
**How to categorize this issue?**

/area control-plane
/kind enhancement
/priority 3

**What would you like to be added**:
`docker` is still used in Gardener internals:
* The `systemd` unit responsible for downloading the cloud-config depends on the `docker` systemd unit: https://github.com/gardener/gardener/blob/2788f29310bb23faa224a72bbf946f12136e0f5f/pkg/operation/botanist/component/extensions/operatingsystemconfig/downloader/downloader.go#L138-L140
  * This probably makes sense, as the unit runs the script below, which uses `docker`
* Gardener preloads the hypercube image and uses `docker` to extract the binaries for kubelet and kubectl: https://github.com/gardener/gardener/blob/2788f29310bb23faa224a72bbf946f12136e0f5f/pkg/operation/botanist/component/extensions/operatingsystemconfig/executor/templates/scripts/execute-cloud-config.tpl.sh#L80-L106
* Promtail uses a similar mechanism like above to extract the binary from an image: https://github.com/gardener/gardener/blob/2788f29310bb23faa224a72bbf946f12136e0f5f/pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/promtail/component.go#L60
* The root certificate updater restarts the docker daemon after updating the certs: https://github.com/gardener/gardener/blob/2788f29310bb23faa224a72bbf946f12136e0f5f/pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/rootcertificates/component.go#L62
* Gardener tunes some kernel parameters for docker-specific reasons: https://github.com/gardener/gardener/blob/2788f29310bb23faa224a72bbf946f12136e0f5f/pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/kernelconfig/component.go#L89-L97

We should remove any usage of `docker` in Gardener, such that Gardener can use OS images without docker daemon installed.

**Proposal**:
* For downloading/preloading images and extracting binaries, we can replace `docker` with an alternative CLI which uses `containerd` instead: `ctr ` should be installed on the OSes supporting `containerd` already today. Alternatively, there is `nerdctl`, which offers a UX similar to `docker`, but supports additional stuff like lazy-pulling (which might come in handy specifically for extracting single binaries from an image: https://github.com/containerd/nerdctl/blob/master/docs/stargz.md)
* Drop the daemon reload after updating root certificates
* Investigate if kernel parameter tuning is still necessary/valid when using `containerd`. We might be able to remove this entirely or need very different optimizations

**Additional considerations**:
Before creating an OS image without support for `docker`, there's a couple of other things to consider:
* Remove the migration step in the CloudProfile controller, adding `docker` into the list of supported runtimes for each image: https://github.com/gardener/gardener/blob/a82aed23505004b507796b9f47a12f19b98d8fe6/pkg/controllermanager/controller/cloudprofile/cloudprofile_control.go#L122-L132
* Remove the validation in https://github.com/gardener/gardener/blob/6f878c51682ba75f8c6281b9f68d5a37998b2b8d/pkg/apis/core/validation/cloudprofile.go#L311-L313, added with https://github.com/gardener/gardener/pull/6480. 
* TODO: more considerations here

**Why is this needed**:
While it is possible for users to specify `containerd` to run their workload containers, Gardener still requires the docker daemon to be present and working in the OS images used for the worker nodes.

**Comment by timebertt** [comment]
`ctr images mount` seems like a missing puzzle piece in this story: https://github.com/containerd/containerd/pull/4210

**Comment by rfranzke** [comment]
I guess this issue can be closed when the new [`gardener-node-agent`](https://github.com/gardener-community/hackathon/tree/main/2023-05_Leverkusen#%EF%B8%8F-introduction-of-gardener-node-agent) is getting introduced. cc @timebertt @majst01 @Gerrit91

**Comment by rfranzke** [comment]
> I guess this issue can be closed when the new [`gardener-node-agent`](https://github.com/gardener-community/hackathon/tree/main/2023-05_Leverkusen#%EF%B8%8F-introduction-of-gardener-node-agent) is getting introduced. cc @timebertt @majst01 @Gerrit91

Actually, it seems that there is another usage of `docker cp`: https://github.com/gardener/gardener/blob/master/pkg/component/extensions/operatingsystemconfig/original/components/valitail/component.go#L63-L65

**Comment by rfranzke** [comment]
With https://github.com/gardener/gardener/pull/8707, we have adapted the `extensions.gardener.cloud/v1alpha1.OperatingSystemConfig` API as follows:

- `.spec.units` now has a new field `files[]` of type `[]extensionsv1alpha1.File`
- `extensionsv1alpha1.FileContent` allows to configure an image reference:
   ```yaml
   spec:
     units:
     - name: kubelet.service
        files:
        - path: /opt/bin/kubelet
           permissions: 0755
           content:
             imageRef:
               image: eu.gcr.io/gardener-project/hyperkube:v1.28.1
               filePathInImage: /kubelet
   ```

`gardener-node-agent` will take care of extracting such binaries from images and copy them to the host (via https://github.com/gardener/gardener/blob/master/pkg/nodeagent/registry/containerd_extractor.go), and then restart the affected units.

This approach will handle:
- `kubelet` binary for the `kubelet.service` unit
- `kubectl` binary for the `kubelet-monitor.service` unit (and human operations tooling)
- `valitail` binary for the `valitail.service` unit ([ref](https://github.com/gardener/gardener/issues/4673#issuecomment-1573925561))
- `gardener-node-agent` binary for the `gardener-node-agent.service` unit (GNA performs a "self-upgrade")

In the `gardener-node-init` script (used for bootstrapping the `gardener-node-agent` on initial VM/node provisioning), we plan to use `ctr` as suggested in https://github.com/gardener/gardener/issues/4673#issuecomment-1507003507 (we already expect it to be present on nodes with `containerd` since a long time as documented in https://github.com/gardener/gardener/blob/master/docs/extensions/operatingsystemconfig.md#contract-operatingsystemconfig-resource).

## TL;DR

With above approaches, we get rid of all usages of `docker cp` and can close this issue as soon as #8023 is closed (as soon as `gardener-node-agent` is GA).

**Comment by majst01** [comment]
What a achievement !

**Comment by rfranzke** [comment]
> What a achievement !

This is massive. Thanks everybody for contributing!
